### ISSUE TITLE
Constructive Critique and Improvement Suggestions for Basalt's Frontend APIs

### ISSUE THREAD

**Issue by obadakhalili** [basalt-org/basalt#50]
I have tried to build a simple model with Basalt to get a sense of its frontend APIs and how it works and I have a few comments I would like to share.

It is a simple linear regression model that has the following code:
```mojo
from basalt import Graph, Tensor, TensorShape, nn, dtype
from basalt.utils import tensorutils
from random import randn


fn create_model_graph() -> Graph:
    var g = Graph()

    var x = g.input(TensorShape(800, 2))
    var y_pred = nn.Linear(g, x, 1)
    g.out(y_pred)
    var y_true = g.input(TensorShape(800, 1))
    var loss = nn.MSELoss(g, y_pred, y_true)
    g.loss(loss)

    return g


fn main() raises:
    alias graph = create_model_graph()
    var model = nn.Model[graph]()
    var optim = nn.optim.Adam[graph](lr=0.01)
    optim.allocate_rms_and_momentum(model.parameters)

    var X_train_data = DTypePointer[dtype].alloc(800 * 2)
    randn(X_train_data, 800 * 2)
    var X_train = Tensor[dtype](X_train_data, TensorShape(800, 2))

    var true_weights = Tensor[dtype](TensorShape(2, 1))
    true_weights[0] = 2
    true_weights[1] = 3

    var y_train = Tensor[dtype](TensorShape(800, 1))
    tensorutils.dot[TensorShape(800, 2), TensorShape(2, 1)](
        y_train, X_train, true_weights
    )

    var X_test_data = DTypePointer[dtype].alloc(200 * 2)
    randn(X_test_data, 200 * 2)
    var X_test = Tensor[dtype](X_test_data, TensorShape(200, 2))

    var y_test = Tensor[dtype](TensorShape(200, 1))
    tensorutils.dot[TensorShape(200, 2), TensorShape(2, 1)](
        y_test, X_test, true_weights
    )

    for epoch in range(1000):
        var loss = model.forward(X_train, y_train)[0]
        optim.zero_grad(model.parameters)
        model.backward()
        optim.step(model.parameters)
        if epoch == 0 or (epoch + 1) % 100 == 0:
            print("Epoch: ", epoch + 1, ", Loss: ", loss, sep="")

    var test_loss = model.forward(X_test, y_test)[0]
    print("Test Loss: ", test_loss, sep="")
    print("Params: ", model.parameters.params.data[2])

```

My main point is that the frontend has a bad user experience. It isn't clear what the `Graph` API is doing at first glance, and once you realize it is for describing your model in some sort of DSL before populating it, it doesn't get easier to use it. It isn't intuitive that you have to describe your model's graph in terms of `input`, `out`, `loss`, `param`, etc. nodes, and the APIs for specifying them aren't great either.
Another pain point is the APIs for initializing tensors. I think it is clear from my code where the problem is. Most of the `main` function's body went for initializing simple train and test tensors.

I think the frontend can be a lot simpler. A good north star to work towards is Pytorch's frontend. tinygrad is an example of a recent framework that is applying this successfully. The result is a simple and intuitive frontend that users are already familiar with or is easy to familiarize themselves with. Here is the code for the same model in Python Pytorch:

```py
import torch

X = torch.randn(1000, 2)
y = X.mv(torch.tensor([2, 3], dtype=torch.float))
X_train = X[:800]
y_train = y[:800]
X_test = X[800:]
y_test = y[800:]

class LinearRegression(torch.nn.Module):
    def __init__(self):
        super(LinearRegression, self).__init__()
        self.linear = torch.nn.Linear(2, 1)

    def forward(self, x):
        return self.linear(x)

model = LinearRegression()
criterion = torch.nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

for epoch in range(1000):
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs.squeeze(), y_train)
    loss.backward()
    optimizer.step()
    if epoch == 0 or (epoch + 1) % 100 == 0:
        print(f'Epoch {epoch + 1}, Loss: {loss.item()}')

with torch.no_grad():
    y_pred = model(X_test)
    loss = criterion(y_pred.squeeze(), y_test)
    print(f'Test loss: {loss.item()}')
    print(f'Model params: {model.linear.weight[0].numpy()} vs True weights: [2, 3]')
```

I'm new to Mojo so I don't know how easy is this to pull off. However, I think this is necessary if the project hopes to compete with existing alternatives (it is not enough that the code is written in Mojo) and it seems that the current frontend is far from the simplest form it can take.

These are my two cents, I hope you accept them ..

**Comment by Benny-Nottonson** [comment]
These are all great and very valid suggestions, I can't give explicit comments as to the direction we will choose, but we are actively looking into this and have already made serious changes since this issue was created. Because of that, im going to close this as the API surrounding Graph has changed quite a bit, but feel free to reopen with more suggestions in the future.

**Comment by obadakhalili** [comment]
Thanks. Glad I could help.
